Support MySQL-based FunctionNamespaceManager

caithagoras opened the pull request:
```
== RELEASE NOTES ==

General Changes
* Add support for creating, altering, and dropping SQL functions with a MySQL-based storage. (:doc:`/admin/function-namespace-managers`)
```

rongrong commented:
A very high level comment: I see that you are renaming the module `presto-sql-function` to `presto-function-namespace-managers` and adding the function namespace manager directly to this module. If you are planning to do this, you probably want to introduce subdirectories to organize the code rather than keep them all under `.../sql-function`.

caithagoras commented:
@rongrong Rename package `sqlfunction` to `functionNamespace`

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,240 @@
+})
+public interface FunctionNamespaceDao
+{
+    @SqlUpdate("CREATE TABLE IF NOT EXISTS <table_name> (\n" +
```
Maybe name it `<function_namespaces_table_name>` to make the query clearer to read.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,240 @@
+            ")")
+    void createFunctionNamespacesTableIfNotExists(@Define("table_name") String tableName);
+
+    @SqlUpdate("CREATE TABLE IF NOT EXISTS <table_name> (\n" +
```
Maybe name this `<sql_invoked_functions_table_name>`.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,240 @@
+            "  catalog_name varchar(128) NOT NULL,\n" +
+            "  schema_name varchar(128) NOT NULL,\n" +
+            "  function_name varchar(256) NOT NULL,\n" +
+            "  parameters text NOT NULL,\n" +
```
Does this still need to be text? Due to `function_id` this has an upper bound in length as well right?

rongrong replied:
MySQL's implementation on `varchar` is more efficient than `text`. `text != varchar(65535)` https://stackoverflow.com/questions/7755629/varchar255-vs-tinytext-tinyblob-and-varchar65535-vs-blob-text

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,240 @@
+            "  parameters text NOT NULL,\n" +
+            "  return_type varchar(256) NOT NULL,\n" +
+            "  language varchar(64) NOT NULL,\n" +
+            "  `deterministic` tinyint(1) NOT NULL,\n" +
```
I think it's more descriptive to use `boolean` rather than tinyint(1). Tinyint(1) is how MySQL implements boolean but that's its implementation detail.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,240 @@
+            "  create_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,\n" +
+            "  update_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,\n" +
+            "  PRIMARY KEY (id),\n" +
+            "  UNIQUE KEY function_id_version (function_id,version),\n" +
```
space between column names.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,240 @@
+            "  update_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,\n" +
+            "  PRIMARY KEY (id),\n" +
+            "  UNIQUE KEY function_id_version (function_id,version),\n" +
+            "  KEY qualified_function_name (catalog_name,schema_name,function_name)\n" +
```
same here.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,240 @@
+
+    @SqlQuery("SELECT\n" +
+            "   count(1) > 0\n" +
+            "FROM <table_name>\n" +
```
Maybe change to more meaningful table name.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,240 @@
+            "  parameters text NOT NULL,\n" +
+            "  return_type varchar(256) NOT NULL,\n" +
+            "  language varchar(64) NOT NULL,\n" +
+            "  `deterministic` tinyint(1) NOT NULL,\n" +
```
maybe rename this column so it's not a keyword? This is a little annoying. 😂Don't know what to name it though...

caithagoras replied:
I renamed it to `is_deterministic`. Sounds good?

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,240 @@
+            "WHERE\n" +
+            "    t.function_id = :function_id\n" +
+            "FOR UPDATE")
+    Optional<SqlInvokedFunctionRecord> getLatestRecord(
```
Name it `getLatestRecordForUpdate`? People not reading the implementation would not be aware of the side effect otherwise.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/SqlInvokedFunctionRowMapper.java:
```diff
@@ -0,0 +1,71 @@
+public class SqlInvokedFunctionRowMapper
+        implements RowMapper<SqlInvokedFunction>
+{
+    private static final JsonCodec<List<SqlParameter>> CODEC = JsonCodec.listJsonCodec(SqlParameter.class);
```
Maybe name it `SQL_PARAMETERS_CODEC`?

rongrong commented on presto-function-namespace-managers/src/test/java/com/facebook/presto/functionNamespace/db/TestDbFunctionNamespaceManager.java:
```diff
@@ -0,0 +1,292 @@
+    private Jdbi jdbi;
+    private DbFunctionNamespaceManager functionNamespaceManager;
+
+    public static String getFunctionNamespacesTableName()
```
Do they need to be public? Maybe use `"test_xxx"`?

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/DbFunctionNamespaceManagerConfig.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Config("max-catalog-name-length")
+    public DbFunctionNamespaceManagerConfig setMaxCatalogNameLength(int maxCatalogNameLength)
```
You are not using these when creating the database tables. And as long as you are using `CREATE TABLE IF NOT EXISTS` you cannot really change the upper bound of this. So either restrict the allowed value, or this doesn't really need to be configurable.

rongrong replied:
The Dao is very much MySQL syntax so the argument that people might keep that while using a different database backend is not very convincing to me. In general it's better to only provide flexibility when absolutely needed otherwise you might end up just creating dead code that hard to reason about later. I'd remove these until it's proven necessary. After all this is not hard to add when it's needed. Currently from this PR I don't see any use for this.

caithagoras commented:
@rongrong 
Made a few improvements in the 3rd commit and will merge into the 2nd commit before merging.
- Use annotation to pre-define all `<function-namespaces-table>` and `<sql-functions-table>` without requiring String to be passed as parameters.
- Use module to create `FunctionNamespaceManager` in test, so that the module is tested as well.
- Update some config names and default values.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/DbFunctionNamespaceManagerConfig.java:
```diff
@@ -52,7 +52,7 @@ public String getFunctionNamespacesTableName()
     }
 
-    @Config("function-namespace-table-name")
+    @Config("function-namespaces-table-name")
```
If you decide to do this, make sure you handle release deployment correctly.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -160,6 +160,41 @@ boolean servesFunctionNamespace(
+            "    t.body,\n" +
+            "    t.version,\n" +
+            "    t.deleted\n" +
+            "FROM <table_name> t\n" +
```
Missed this one?

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/AbstractSqlInvokedFunctionNamespaceManager.java:
```diff
@@ -43,19 +46,24 @@
 
     public AbstractSqlInvokedFunctionNamespaceManager(SqlInvokedFunctionNamespaceManagerConfig config)
     {
+        this.servingNamespaces = config.getServingNamespaces().stream()
```
The serving function namespaces should come from the table rather than the config. The config provides initial values. Database has the source of truth. Also, the initial values is supposed to be a stop-gap for proper grant support.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/SqlInvokedFunctionNamespaceManagerConfig.java:
```diff
@@ -14,17 +14,39 @@
     private Duration functionInstanceCacheExpiration = new Duration(8, HOURS);
 
+    @NotNull
+    public List<String> getServingNamespaces()
```
Instead of duplicating the config in two places, maybe change the `listFunctions` api to `listFunctions(prefix)` and collect that from `FunctionManager` directly.

rongrong commented:
Test failure is related.

rongrong commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java:
```diff
@@ -198,15 +205,32 @@ public void registerBuiltInFunctions(List<? extends BuiltInFunction> functions)
+
+                if (!servingNamespaces.contains(functionNamespace.asCatalogSchemaPrefix())
+                        && !servingNamespaces.contains(new CatalogSchemaPrefix(functionNamespace.getCatalogName(), Optional.empty()))) {
+                    throw new PrestoException(
```
Hmm, do we need to through error if the function namespace manager has some old functions from a namespace that's no longer used? Personally I prefer the function namespace manager to support a `listFunction(prefix)` API. This way we simply ignore anything that's in the function namespace manager but not configured to be managed by this instance. (Think, the same backend database hosting function namespaces `a.b`, `x.y`, but for one cluster you are only configuring it to manage `a.b`, while on another cluster it's configured to manage `x.y`). I don't see why this needs to be an internal error. It seems to be reasonable. Essentially, the config is constructing a view of the underlying database to be exposed to a certain cluster configuration.

rongrong commented on presto-main/src/main/java/com/facebook/presto/metadata/StaticFunctionNamespaceStore.java:
```diff
@@ -70,7 +70,7 @@ private void loadFunctionNamespaceManager(File file)
         checkState(functionNamespaceManagerName != null, "Function namespace configuration %s does not contain function-namespace-manager.name", file.getAbsoluteFile());
 
-        String functionNamespaces = properties.remove("serving-namespaces");
+        String functionNamespaces = properties.get("serving-namespaces");
```
Is this still necessary? Where is this used other than here?

caithagoras commented:
@rongrong 
- Squashed `DbFunctionNamespaceManager` changes into 1 commit.

3 new commits:
  - Require function namespace manager to bind with catalog names
  - "Will merge into previous commit"
  - Add `presto-docs`.

caithagoras commented:
Per offline discussion, we decided to move function namespace a bit closer to the connector model.
- Each function namespace serves one or more "catalogs", but it cannot only serve a subset of `schemas` within a `catalog`.
- `FunctionNamespaceManagerFactory.create` takes the set of catalogs to construct the `FunctionNamespaceManager`.
- `FunctionManager` manages the mapping between `catalog` to `FunctionNamespaceManager`. (catalog can be thought as the `ConnectorId`, which is essentially a wrapper around `catalogName`).

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/DbFunctionNamespaceManagerConfig.java:
```diff
@@ -52,25 +39,6 @@ public DbFunctionNamespaceManagerConfig setFunctionNamespacesTableName(String fu
     }
 
-    @NotNull
-    public List<CatalogSchemaName> getDefaultFunctionNamespaces()
```
So we will not provide a convenient initialization from config files anymore? People have to manually insert values to the db?

caithagoras replied:
I don't think this configuration property is that necessary since admin can just insert values into their DBs. It is also not an exactly a good design since it would be better to create function namespaces (whether the initial ones or not) at system level so it goes through proper code paths and checks, instead of in each "connector's" configuration. What just described will be supported by the `CREATE SCHEMA` statement.

rongrong replied:
Make sure this is documented then.

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,34 @@
+functions with a persistent storage, allowing users to create, alter,
+and delete functions, as well as to reference them within queries.
+
+Presto by default supports a MySQL-based function namespace manager.
```
`Presto provides a MySQL-based ... ` I don't think it's enabled "by default". You need to configure it to use it.

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,34 @@
+It is pluggable and can be configured by an administrator.
+
+To instantiate a MySQL-based function namespace manager with id
+``prod``, add a file ``etc/function-namespace/prod.properties``
```
Let's use maybe "test" or "example" rather than the word "prod". Maybe double check what other examples are using.

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,34 @@
+``prod``, add a file ``etc/function-namespace/prod.properties``
+with the following contents::
+
+    function-namespace-manager.name=xdb
```
xdb?

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,34 @@
+with the following contents::
+
+    function-namespace-manager.name=xdb
+    serving-catalogs=catalog1.schema1,catalog2
```
I thought this is only configuring catalog, how does it work with `catalog1.schema1`?

rongrong replied:
If you are planning to go with each function namespace manager will manage full catalog, just use the same model as connectors where the name of the file is the catalog name, and each config will manage a single catalog. Otherwise the naming is really confusing. Does the name of the config file mean anything?

caithagoras replied:
Moving to 1 catalog per FNM model, with the file name being the catalog name.

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,34 @@
+=========================================== ================================================================================================
+Name                                        Description
+=========================================== ================================================================================================
+``function-namespace-manager.name``         The type of the function namespace manager to instantiate. Currently, only ``xdb`` is supported.
```
I think you want to use `db` or `mysql`

caithagoras commented:
@rongrong Comments addressed. Thanks for revieweing!

rongrong commented:
A few high level thoughts:
* You might want to put only allow one catalog per config and removing `default-function-namespaces` into separate commits. These are not related to the introduction of mysql-based function namespace manager. They should be refactor commits before the function namespace manager implementation.
* I'm thinking, what about put all routine characteristics into one column and use text/JSON to store it? We don't seem to ever filter on these columns in db queries anyways. And we might add more routine characteristics as we extend function support.

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,50 @@
+    database_url=localhost:1080
+
+When Presto first starts with the above MySQL function namespace
+manager configuration, two MySQL tables are created.
```
two MySQL tables *will be* created?

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,50 @@
+add a file ``etc/function-namespace/example.properties`` with the
+following contents::
+
+    function-namespace-manager.name=mysql
```
I think it's clearer to add the function namespace and functions table here in the config. That way later you can describe what tables are created better.

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,50 @@
+When Presto first starts with the above MySQL function namespace
+manager configuration, two MySQL tables are created.
+
+- ``function_namespaces``: Stores all function namespaces.
```
Use the configured table name from the example config instead.

The concept of "function namespace" was never officially introduced and defined. Please give a better explanation. What does "all function namespaces" mean? Maybe "allowed function namespaces"?

rongrong commented on presto-docs/src/main/sphinx/sql/alter-function.rst:
```diff
@@ -0,0 +1,47 @@
+==============
```
@kaikalur Can you take a look at the documentation for create / alter / drop function? Thanks!

kaikalur commented on presto-docs/src/main/sphinx/sql/alter-function.rst:
```diff
@@ -0,0 +1,47 @@
+
+Each ``ALTER FUNCTION`` statement can only modify one function
+at a time. If multiple functions are matched by not specifying
+the parameter type list, the query would fail.
```
So we support overloading?!

caithagoras replied:
We don't. Here is my re-words.
```
Parameter type list must be specified if there are multiple
signatures for the specified function name. Otherwise, the query
would fail.

If exactly one signature exists for the specified function name,
parameter type list might be omitted.
```

kaikalur commented on presto-docs/src/main/sphinx/sql/create-function.rst:
```diff
@@ -0,0 +1,83 @@
+:doc:`/admin/function-namespace-managers`.
+
+The optional ``OR REPLACE`` clause causes the ``ALREADY_EXISTS`` error
+to be suppressed if the function already exists.
```
I will simply say something like - 'if a function with identical signature exists, this will quietly replace that'. It's not just the error message.

kaikalur commented on presto-docs/src/main/sphinx/sql/create-function.rst:
```diff
@@ -0,0 +1,83 @@
+to be suppressed if the function already exists.
+
+The ``return_type`` needs to match the actual type of the routine body
+``expression``, without performing type coercion.
```
Match or assignment-compatible? For example, I can return 0 for a double returning function?

caithagoras replied:
This is not supported currently, but we're looking to support it in the future.

kaikalur commented on presto-docs/src/main/sphinx/sql/create-function.rst:
```diff
@@ -0,0 +1,83 @@
+``expression``, without performing type coercion.
+
+A set of routine characteristics can be specified to decorate the
+function and modify its behavior. Each kind of routine
```
modify -> specify

kaikalur commented on presto-docs/src/main/sphinx/sql/create-function.rst:
```diff
@@ -0,0 +1,83 @@
+============================ ======================== ================================================================
+Language clause              SQL                      The language in which the function is defined.
+Deterministic characteristic NOT DETERMINISTIC        Whether the function is deterministic. ``NOT DETERMINISTIC``
+                                                      means that the function is possibly non-deterministic.
```
Be more explicit - specifies that the function is not deterministic. Also, if there is a default, mention it.

caithagoras replied:
Also, according to SQL Spec, `DETERMINISTIC` means the function is guaranteed deterministic, while `NOT DETERMINISTIC` means the functions is "possibly non-deterministic", that is, non-deterministic for certain inputs/usages. `possibly non-deterministic` sounds a more rigorous expression.

The second column shows the default value.

kaikalur commented on presto-docs/src/main/sphinx/sql/create-function.rst:
```diff
@@ -0,0 +1,83 @@
+Deterministic characteristic NOT DETERMINISTIC        Whether the function is deterministic. ``NOT DETERMINISTIC``
+                                                      means that the function is possibly non-deterministic.
+Null-call clause             CALLED ON NULL INPUT     The behavior of the function in which ``null`` is supplied as
+                                                      the value of at least one argument.
```
Similarly here - please specify default

caithagoras replied:
The second column shows the default value.

caithagoras commented:
Changes since last review:
- Change the routine body of `CREATE FUNCTION` statement and `spi/SqlInvokedFunction` to be a Return statement. The SQL text including the `RETURN` keyword is now going to be stored, instead of only the expression of the return statement. This allows us to easily extend the support for other types of routine body.
- Use a single column to represent all routine characteristics as json.

Since config property `default-function-namespace` was never officially introduced, it's not necessary to create a separate commit.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionNamespaceManagerFactory.java:
```diff
@@ -21,5 +21,5 @@
     FunctionHandleResolver getHandleResolver();
 
-    FunctionNamespaceManager<?> create(Map<String, String> config);
+    FunctionNamespaceManager<?> create(String catalog, Map<String, String> config);
```
`catalogName` as well?

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/AbstractSqlInvokedFunctionNamespaceManager.java:
```diff
@@ -43,19 +45,22 @@
     private final LoadingCache<SqlFunctionHandle, ScalarFunctionImplementation> implementationByHandle;
 
-    public AbstractSqlInvokedFunctionNamespaceManager(SqlInvokedFunctionNamespaceManagerConfig config)
+    public AbstractSqlInvokedFunctionNamespaceManager(String catalogName, SqlInvokedFunctionNamespaceManagerConfig config)
```
This seems to be more related to the catalog config commit. But it's fine to keep it here as well.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/DbFunctionNamespaceManager.java:
```diff
@@ -0,0 +1,285 @@
+        checkCatalog(functionHandle);
+        Optional<SqlInvokedFunction> function = functionNamespaceDao.getFunction(functionHandle.getFunctionId(), functionHandle.getVersion());
+        if (!function.isPresent()) {
+            throw new InvalidFunctionHandleException(functionHandle);
```
Do you want to add an optional message as well?

caithagoras replied:
`InvalidFunctionHandleException` adds a error message itself.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/testing/SqlInvokedFunctionTestUtils.java:
```diff
@@ -34,7 +34,11 @@ private SqlInvokedFunctionTestUtils()
     }
 
-    public static final QualifiedFunctionName POWER_TOWER = QualifiedFunctionName.of(new CatalogSchemaName("unittest", "memory"), "power_tower");
+    public static final String TEST_CATALOG = "test";
```
Why not using "unittest" to keep it consistent?

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,295 @@
+            "    t.description,\n" +
+            "    t.language,\n" +
+            "    t.is_deterministic,\n" +
+            "    t.called_on_null_input,\n" +
```
Are you planning to merge these?

caithagoras commented:
Addressed comments about package structure. Merged package `db` and `mysql` into `mysql`.
- Renamed package `functionNamespace.db` to `functionNamesapce.mysql`.
- Renamed classes prefixed with `DbFunctionNamespace` to `MySqlFunctionNamespace`
- Renamed the original `MySqlFunctionNamespaceManagerConfig` to `MySqlConnectionConfig` as it is only about connecting to a MySQL DB.
- Renamed the original `MySqlFunctionNamespaceManagerModule` to `MySqlConnectionModule` for the similar reason.

caithagoras commented:
Hi @sachdevs , could you help taking a look the documentation in the last commit, and see if it is easy for you to understand? The doc is intended for Presto devs and admins who do not have priori knowledge about function namespace.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/testing/SqlInvokedFunctionTestUtils.java:
```diff
@@ -44,7 +44,7 @@ private SqlInvokedFunctionTestUtils()
             "power tower",
             RoutineCharacteristics.builder().setDeterminism(DETERMINISTIC).build(),
-            "pow(x, x)",
+            "RETURN pow(x, x)",
```
This would cause compatibility issue when roll out to production. Make sure that's handled during deployment.

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,76 @@
+------------------
+
+A function namespace is a catalog-schema name (i.e. ``example.test``)
+in which each function can be uniquely identified by its signature,
```
signature includes return type and other information as well. So just say "can be uniquely identified by its name and parameter type list".

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,76 @@
+in which each function can be uniquely identified by its signature,
+that is, its name and parameter type list.
+
+The function namespace, such as ``example.test``, is not a full-fledged
```
"A function namespace"? What's the behavior if I use a real catalog as function catalog? I think it will still work right? So potentially "example.text" can be a real catalog.schema as well.

caithagoras replied:
I think we should disallow registering function namespace with a catalog that has already been registered with a connector. This will cause lots of confusion as the sets of schema will be different from the sets of function namespaces within that catalog, and makes support `CREATE SCHEMA`, `DROP SCHEMA`, and `SHOW SCHEMA` on function namespace managed catalogs problematic or even infeasible.

caithagoras replied:
If we have connector functions one day, we can reverse that restriction and properly support that use case.

rongrong replied:
But we actually don't disallow it in code right? So call this out specifically. Something like using existing connector catalog is not supported and the behavior is not defined or tested.

caithagoras replied:
```
Each instance of function namespace manager binds to a catalog name,
as it manages all functions within that catalog. Using the catalog
name of an existing connector is discouraged, as the behavior is not
defined nor tested, and will be disallowed in the future.

Currently, those catalog names do not corresponds to real catalogs
as they cannot be specified as the catalog in session, nor do they
support :doc:`CREATE SCHEMA`, :doc:`ALTER SCHEMA`, :doc:`DROP SCHEMA`,
or :doc:`SHOW SCHEMAS`. Instead, namespaces can be added using the
methods described below.
```

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,76 @@
+schema as it does not support storing tables and views, but only
+functions.
+
+Each instance of function namespace manager manages a catalog. Creating
```
What is a "function namespace manager"? Define it first.

caithagoras replied:
It purpose and functionality is already defined in the first paragraph. Here, I'm expanding the description.

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,76 @@
+Configuration
+-------------
+
+Presto provides a MySQL-based function namespace manager. It is
```
Maybe "Presto provides a MySQL-based function namespace manager *implementation*".

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,76 @@
+the same set of tables, or different tables / MySQL databases.
+
+To create a new function namespace, insert into the
+``function_namespaces`` table::
```
I think you mean ``example_function_namespaces`` table here.

rschlussel commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,76 @@
+-------------
+
+Presto provides a MySQL-based function namespace manager. It is
+pluggable and can be configured by an administrator.
```
What does pluggable mean here? Also, how else would it be configured if not by an administrator?

rschlussel commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,76 @@
+To create a new function namespace, insert into the
+``function_namespaces`` table::
+
+    INSERT INTO function_namespaces (catalog_name, schema_name)
```
should this table be `example_function_namespaces`?  We haven't mentioned the `function_namespaces` table yet

rschlussel commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,76 @@
+schema as it does not support storing tables and views, but only
+functions.
+
+Each instance of function namespace manager manages a catalog. Creating
```
what do you mean that `Creating and managing schema in those catalogs are not supported`?  Don't you describe below how to do that?

caithagoras replied:
I meant you cannot use `CREATE SCHEMA`, `DROP SCHEMA`, `SHOW SCHEMAS` on a catalog managed by a function namespace manager.

sachdevs replied:
Let's be explicit about that since that was not obvious to me as well.

rschlussel replied:
Got it.  Maybe say it more directly then.  Something like:
CREATE SCHEAM, DROP SCHEMA, and SHOW SCHEMAS is not currently supported on catalogs managed by the function namespace manager.  Instead, namespaces can be added using the syntax described below.

rschlussel commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,76 @@
+If multiple function namespace managers are instantiated, they may use
+the same set of tables, or different tables / MySQL databases.
+
+To create a new function namespace, insert into the
```
I'm a bit confused about this.  
1. If each function namespace manager only manages a single catalog, then why does the function_namespaces catalog take in both a catalog and schema name. 
2. Above you said you can't add schemas

caithagoras replied:
For 1: Different function namespace managers may use the same `function_namespaces`, and hence the table may contain function namespaces not managed by a single function namespace manager.

For 2:
a. You cannot create a function namespace with `CREATE SCHEMA`.
b. A function namespace is not exactly the same as a schema, i.e., you cannot create tables and views in it.
Although I agree it might just be ok to call it a schema when we support (a).

rschlussel replied:
2 makes sense.  

For 1 are you saying that different function namespace managers may use the same table to keep track of their namespaces?  It would be helpful to state that explicitly somewhere.   Also, do you need to have separate configuration files for each catalog, or is it enough to stick new info in the function_namespaces table?

caithagoras replied:
The section above states it:
```
If multiple function namespace managers are instantiated, they may use
the same set of tables, or different tables / MySQL databases.
```

caithagoras replied:
I expanded the explanation for multiple function namespace manager.
```
Multiple function namespace managers can be instantiated by placing
configuration properties file under ``etc/function-namespace``. They
may use the same tables, or different tables / MySQL databases. If the
same MySQL table are used by multiple function namespace managers,
each manager will only create and interact with entries of the catalog
to which it binds.
```

rschlussel commented on presto-docs/src/main/sphinx/sql/alter-function.rst:
```diff
@@ -0,0 +1,49 @@
+for the specified function name. Otherwise, the query would fail.
+
+If exactly one signature exists for the specified function name,
+parameter type list might be omitted.
```
nit: might -> can

rschlussel commented on presto-docs/src/main/sphinx/sql/alter-function.rst:
```diff
@@ -0,0 +1,49 @@
+Alter the definition of an existing function.
+
+Parameter type list must be specified if multiple signatures exist
+for the specified function name. Otherwise, the query would fail.
```
nit: remove `Otherwise the query would fail`

rschlussel commented on presto-docs/src/main/sphinx/sql/alter-function.rst:
```diff
@@ -0,0 +1,49 @@
+If exactly one signature exists for the specified function name,
+parameter type list might be omitted.
+
+A subset of routine characteristics, as defined in
```
Maybe say that right now only modifying the null-call clause is supported.  Otherwise it's pretty surprising as a user to get to the end of this documentation and discover that mostly you can't modify the function.

rschlussel commented on presto-docs/src/main/sphinx/sql/create-function.rst:
```diff
@@ -0,0 +1,84 @@
+characteristic can be specified at most once.
+
+============================ ======================== ================================================================
+Routine Characteristic       Default Value            Description
```
For all of these, what other (non-default) options are supported

caithagoras replied:
The other values are available in the `Synopsis` section. Placing them in the table may jeopardize readability.

rschlussel commented on presto-docs/src/main/sphinx/sql/drop-function.rst:
```diff
@@ -0,0 +1,43 @@
+
+Drop an existing function.
+
+The optional ``OR REPLACE`` clause causes the ``NOT_FOUND`` error
```
i think you mean ``IF EXISTS`` clause

sachdevs commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,78 @@
+Function Namespace
+------------------
+
+A function namespace is a catalog-schema name (i.e. ``example.test``)
```
Nit: I think you mean e.g. and not I.e.

sachdevs replied:
Instead of saying catalog-schema it may make more sense to use the same syntax as your example to be explicit like catalog.schema

sachdevs commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,78 @@
+in which each function can be uniquely identified by its name and
+parameter type list.
+
+A function namespace, such as ``example.test``, is not a full-fledged
```
I had to read this a couple times to clarify. Maybe say that: `a function namespace can be thought of as a schema for storing functions. However, it is not a full fledged schema as it does not support...`.

sachdevs commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,78 @@
+- ``example_sql_functions`` stores SQL-invoked functions of the
+  catalog ``example``.
+
+If multiple function namespace managers are instantiated, they may use
```
This was unclear to me. Does this simply mean that multiple function namespace managers can be configured to use the same tables?

caithagoras replied:
Yes. Rewrote into
```
Multiple function namespace managers can be instantiated by placing
multiple properties files under ``etc/function-namespace``. They
may be configured to use the same tables. If so, each manager will
only create and interact with entries of the catalog to which it binds.
```

rongrong commented:
The test failures seem to be related.

caithagoras commented:
All comments addressed. Thanks for the feedback!

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,88 @@
+
+Function namespace managers support storing and retrieving SQL
+functions, allowing Presto engine to perform actions such as creating,
+altering, deleting, and executing functions.
```
"executing functions" is not new and it's not something "allowed" by function namespace managers.

caithagoras replied:
I intend to say `resolving functions`

rongrong replied:
resolving function is also not introduced by function namespace manager though.

rongrong commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,88 @@
+functions. However, it is not a full fledged schema as it does not
+support storing tables and views, but only functions.
+
+Each Presto function, whether built-in or user-defined, resides in
```
Introduce the concept of "qualified function name first". `A qualified function name is a .... Each Presto function, whether ......`

rongrong replied:
Should we *Italy* the concepts like *function namespace manager*, *function namespace*, *qualified function name*? What's the convention?

caithagoras replied:
I don't see those kinds of formatting be used in presto-docs.

caithagoras commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/DbFunctionNamespaceManagerConfig.java:
```diff
@@ -0,0 +1,67 @@
+{
+    private String functionNamespacesTableName = "function_namespaces";
+    private String functionsTableName = "sql_functions";
+    private Duration queryTimeout = new Duration(20, SECONDS);
```
I believe the default is no timeout at all if not specified explicitly.

sachdevs commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,93 @@
+name. A function is uniquely identified by its qualified function name
+and parameter type list.
+
+Each instance of function namespace manager binds to a catalog name
```
of `a` function namespace manager

caithagoras replied:
When we're saying `a function namespace manager`, we're referencing an instance, not a type of function namespace manager. So, maybe, I should just each `Each function namespace manager...`.

sachdevs commented on presto-docs/src/main/sphinx/admin/function-namespace-managers.rst:
```diff
@@ -0,0 +1,93 @@
+Configuration
+-------------
+
+Presto provides a MySQL-based function namespace manager
```
I originally did not understand what MySQL-based implementation meant at first (until I read the release note). Maybe reword this to `Presto currently stores all function namespace manager related information in MySQL.`

sachdevs commented on presto-docs/src/main/sphinx/sql/create-function.rst:
```diff
@@ -0,0 +1,84 @@
+and its parameter type list. ``qualified_function_name`` needs to be in
+the format of ``catalog.schema.function_name``.
+
+To create a function, the function namespace as identified by
```
`In order to create a function, the corresponding function namespace (in the format catalog.schema) must first be managed by a function namespace manager`

rongrong commented:
For the release note, I'd mention "function namespace manager" and refer to the documentation added.
